zkevm-circuits/src/util.rs

```diff
@@ -212,3 +217,19 @@ pub(crate) fn get_push_size(byte: u8) -> u64 {
         0u64
     }
 }
+
+/// Returns number of unusable rows of the Circuit.
+/// The minimum unusable rows of a circuit is currently 6, where
+/// - 3 comes from minimum number of distinct queries to permutation argument witness column
+/// - 1 comes from queries at x_3 during multiopen
+/// - 1 comes as slight defense against off-by-one errors
+/// - 1 comes from reservation for last row of permutation argument witness column
+///
+/// For circuit with column queried at more than 3 distinct rotation, we can
+/// calculate the unusable rows as (x - 3) + 6 where x is the number of distinct
+/// qureies.
+pub(crate) fn unusable_rows<F: Field, C: Circuit<F>>() -> usize {
+    let mut cs = ConstraintSystem::default();
+    C::configure(&mut cs);
+    cs.blinding_factors() + 1 // +1 for l_last
```

Comment:
It's internal to halo2, it's already in upstream actually, so it's because we need to reserve an extra row for grand-product boundray check, so that row is not copy-able or lookup-able, I have added more comment about this extra one in https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1282/commits/d54f0564ef955d5bf9c26407be80148fd53fa329